File: ios/HybridNitroText.swift
//
//  HybridNitroText.swift
//  Pods
//
//  Created by Patrick Kabwe on 9/1/2025.
//

import Foundation
import UIKit

class HybridNitroText: HybridNitroTextSpec, NitroTextViewDelegate {
    private let textView = NitroTextView()
    var view: UIView { textView }
    let nitroTextImpl: NitroTextImpl

    override init() {
        self.nitroTextImpl = NitroTextImpl(textView)
        super.init()
        self.textView.nitroTextDelegate = self
    }

    func onNitroTextLayout(_ layout: TextLayoutEvent) {
        onTextLayout?(layout)
    }

    func onNitroTextPressIn() { onPressIn?() }
    func onNitroTextPressOut() { onPressOut?() }
    func onNitroTextPress() { onPress?() }

    // Props

    var fragments: [Fragment]? {
        didSet {
            nitroTextImpl.setFragments(fragments)
        }
    }

    var renderer: NitroRenderer? {
        didSet {
            nitroTextImpl.setRenderer(renderer)
        }
    }

    var richTextStyleRules: [RichTextStyleRule]? {
        didSet {
            nitroTextImpl.setRichTextStyleRules(richTextStyleRules)
        }
    }

    var selectable: Bool? {
        didSet {
            nitroTextImpl.setSelectable(selectable)
        }
    }

    var allowFontScaling: Bool? {
        didSet {
            nitroTextImpl.setAllowFontScaling(allowFontScaling)
        }
    }

    var onTextLayout: ((TextLayoutEvent) -> Void)? = nil
    var onPress: (() -> Void)? = nil
    var onPressIn: (() -> Void)? = nil
    var onPressOut: (() -> Void)? = nil

    var fontSize: Double? {
        didSet {
            nitroTextImpl.setFontSize(fontSize)
        }
    }

    var fontWeight: FontWeight? {
        didSet { nitroTextImpl.setFontWeight(fontWeight) }
    }

    var fontColor: String? {
        didSet { nitroTextImpl.setFontColor(fontColor) }
    }

    var fragmentBackgroundColor: String? {
        didSet { nitroTextImpl.setFragmentBackgroundColor(fragmentBackgroundColor) }
    }

    var fontStyle: FontStyle? {
        didSet { nitroTextImpl.setFontStyle(fontStyle) }
    }

    var fontFamily: String? {
        didSet {
            nitroTextImpl.setFontFamily(fontFamily)
        }
    }

    var textAlign: TextAlign? {
        didSet {
            nitroTextImpl.setTextAlign(textAlign)
        }
    }

    var textTransform: TextTransform? {
        didSet {
            nitroTextImpl.setTextTransform(textTransform)
        }
    }

    var textDecorationLine: TextDecorationLine? {
        didSet { nitroTextImpl.setTextDecorationLine(textDecorationLine) }
    }

    var textDecorationColor: String? {
        didSet { nitroTextImpl.setTextDecorationColor(textDecorationColor) }
    }

    var textDecorationStyle: TextDecorationStyle? {
        didSet { nitroTextImpl.setTextDecorationStyle(textDecorationStyle) }
    }

    var selectionColor: String? {
        didSet {
            nitroTextImpl.setSelectionColor(selectionColor)
        }
    }

    var lineHeight: Double? {
        didSet { nitroTextImpl.setLineHeight(lineHeight) }
    }

    var letterSpacing: Double? {
        didSet { nitroTextImpl.setLetterSpacing(letterSpacing) }
    }

    var text: String? {
        didSet {
            nitroTextImpl.setText(text)
        }
    }

    var numberOfLines: Double? {
        didSet {
            nitroTextImpl.setNumberOfLines(numberOfLines)
        }
    }

    var ellipsizeMode: EllipsizeMode? {
        didSet {
            nitroTextImpl.setEllipsizeMode(ellipsizeMode)
        }
    }

    var dynamicTypeRamp: DynamicTypeRamp? {
        didSet {
            nitroTextImpl.setDynamicTypeRamp(dynamicTypeRamp)
        }
    }

    var lineBreakStrategyIOS: LineBreakStrategyIOS? {
        didSet {
            nitroTextImpl.setLineBreakStrategyIOS(lineBreakStrategyIOS)
        }
    }

    var maxFontSizeMultiplier: Double? {
        didSet {
            nitroTextImpl.setMaxFontSizeMultiplier(maxFontSizeMultiplier)
        }
    }

    var adjustsFontSizeToFit: Bool? {
        didSet {
            nitroTextImpl.setAdjustsFontSizeToFit(adjustsFontSizeToFit)
        }
    }

    var minimumFontScale: Double? {
        didSet {
            nitroTextImpl.setMinimumFontScale(minimumFontScale)
        }
    }

    func afterUpdate() {
        if nitroTextImpl.consumeNeedsCommit() {
            nitroTextImpl.commit()
        }
        textView.setNeedsLayout()
    }
}

File: android/src/main/java/com/nitrotext/HybridNitroText.kt
package com.nitrotext

import androidx.annotation.Keep
import com.facebook.proguard.annotations.DoNotStrip
import com.facebook.react.uimanager.ThemedReactContext
import com.margelo.nitro.nitrotext.*

@Keep
@DoNotStrip
class HybridNitroText(val context: ThemedReactContext) : HybridNitroTextSpec(), NitroTextViewDelegate {
  override val view: NitroTextView = NitroTextView(context)
  private val impl = NitroTextImpl(view.textView)
  private var onTextLayoutCallback: ((TextLayoutEvent) -> Unit)? = null

  init {
    view.nitroTextDelegate = this
  }

  override var fragments: Array<Fragment>?
    get() = null
    set(value) {
      impl.setFragments(value)
    }

  override var selectable: Boolean?
    get() = null
    set(value) {
      impl.setSelectable(value)
    }

  override var allowFontScaling: Boolean?
    get() = null
    set(value) {
      impl.setAllowFontScaling(value)
    }

  override var ellipsizeMode: EllipsizeMode?
    get() = null
    set(value) {
      impl.setEllipsizeMode(value)
    }

  override var numberOfLines: Double?
    get() = null
    set(value) {
      impl.setNumberOfLines(value)
    }

  override var lineBreakStrategyIOS: LineBreakStrategyIOS?
    get() = null
    set(value) {
      // iOS only
    }

  override var dynamicTypeRamp: DynamicTypeRamp?
    get() = null
    set(value) {
      // iOS only
    }

  override var maxFontSizeMultiplier: Double?
    get() = null
    set(value) {
      impl.setMaxFontSizeMultiplier(value)
    }

  override var adjustsFontSizeToFit: Boolean?
    get() = null
    set(value) {
      // TODO: Implement adjustsFontSizeToFit
    }

  override var minimumFontScale: Double?
    get() = null
    set(value) { value }

  override var onTextLayout: ((TextLayoutEvent) -> Unit)?
    get() = onTextLayoutCallback
    set(value) {
      onTextLayoutCallback = value
    }

  override var onPress: (() -> Unit)?
    get() = null
    set(value) {
      // TODO: Implement onPress
    }

  override var onPressIn: (() -> Unit)?
    get() = null
    set(value) {
      // TODO: Implement onPressIn
    }

  override var onPressOut: (() -> Unit)?
    get() = null
    set(value) {
      // TODO: Implement onPressOut
    }

  override var text: String?
    get() = null
    set(value) {
      impl.setText(value)
    }

  override var selectionColor: String?
    get() = null
    set(value) {
      impl.setSelectionColor(value)
    }

  override var fontSize: Double?
    get() = null
    set(value) { impl.setFontSize(value) }

  override var fontWeight: FontWeight?
    get() = null
    set(value) { impl.setFontWeight(value) }

  override var fontColor: String?
    get() = null
    set(value) {
      impl.setFontColor(value)
    }

  override var fragmentBackgroundColor: String?
    get() = null
    set(value) {
      impl.setFragmentBackgroundColor(value)
    }

  override var fontStyle: FontStyle?
    get() = null
    set(value) {  impl.setFontStyle(value) }

  override var fontFamily: String?
    get() = null
    set(value) { impl.setFontFamily(value) }

  override var lineHeight: Double?
    get() = null
    set(value) { impl.setLineHeight(value) }

  override var letterSpacing: Double?
    get() = null
    set(value) { impl.setLetterSpacing(value) }

  override var textAlign: TextAlign?
    get() = null
    set(value) { impl.setTextAlign(value) }

  override var textTransform: TextTransform?
    get() = null
    set(value) { impl.setTextTransform(value) }

  override var textDecorationLine: TextDecorationLine?
    get() = null
    set(value) { impl.setTextDecorationLine(value) }

  override var textDecorationColor: String?
    get() = null
    set(value) {impl.setTextDecorationColor(value) }

  override var textDecorationStyle: TextDecorationStyle?
    get() = null
    set(value) { impl.setTextDecorationStyle(value) }

  override fun afterUpdate() {
    impl.commit()
    view.requestLayout()
    view.invalidate()
  }

  override fun onNitroTextLayout(event: TextLayoutEvent) {
    onTextLayoutCallback?.invoke(event)
  }
}
